File: src/Vec2.cpp
//
// Created by Paolo Broglio on 16/05/25.
//

#include "Vec2.hpp"
#include <iostream>

Vec2::Vec2() : x(0), y(0) {
}

Vec2::Vec2(float x, float y) : x(x), y(y) {
}

float Vec2::getMagnitude() const {
    return sqrt(x * x + y * y);
}

void Vec2::normalize() {
    float magnitude = getMagnitude();
    this->x /= magnitude;
    this->y /= magnitude;
}

Vec2 Vec2::cross(const Vec2 &other) const {
    return  {
        this->x * other.y - this->y * other.x,
        this->y * other.x - this->x * other.y
    };
}

float Vec2::dot(const Vec2 &other) const {
    return this->x * other.x + this->y * other.y;
}

float Vec2::getX() const {
    return this->x;
}

float Vec2::getY() const {
    return this->y;
}

void Vec2::setX(float x) {
    this->x = x;
}

void Vec2::setY(float y) {
    this->y = y;
}

Vec2 Vec2::operator+(const Vec2 &other) const {
    return {x + other.x, y + other.y};
}

Vec2 Vec2::operator-(const Vec2 &other) const {
    return {x - other.x, y - other.y};
}

Vec2 Vec2::operator/(float scalar) const {
    return {x / scalar, y / scalar};
}

Vec2 Vec2::operator*(float scalar) const {
    return {x * scalar, y * scalar};
}

std::ostream & operator<<(std::ostream &os, const Vec2 &vec) {
    os << "Vec2(x=" << vec.getX() << ", y=" << vec.getY() << ")";
    return os;
}

File: src/vec2.zig
pub const Vec2 = struct {
    x: f32,
    y: f32,

    pub fn magnitude(self: *const Vec2) f32 {
        return @sqrt(self.x * self.x + self.y * self.y);
    }

    pub fn normalized(self: *const Vec2) Vec2 {
        const magni: f32 = self.magnitude();
        return Vec2{ .x = self.x / magni, .y = self.y / magni };
    }

    pub fn crossProd(self: *const Vec2, other: Vec2) Vec2 {
        return Vec2{ .x = self.x * other.y - self.y * other.x, .y = self.y * other.x - self.x * other.y };
    }

    pub fn dotProd(self: *const Vec2, other: Vec2) f32 {
        return self.x * other.y + self.y * other.y;
    }

    pub fn add(self: *const Vec2, other: Vec2) Vec2 {
        return Vec2{ .x = self.x + other.x, .y = self.y + other.y };
    }

    pub fn sub(self: *const Vec2, other: Vec2) Vec2 {
        return Vec2{ .x = self.x - other.x, .y = self.y - other.y };
    }

    pub fn scalarProd(self: *const Vec2, n: f32) Vec2 {
        return Vec2{ .x = self.x * n, .y = self.y * n };
    }

    pub fn scalarDiv(self: *const Vec2, n: f32) Vec2 {
        return self.scalarProd(1 / n);
    }
};
